### ISSUE TITLE
bounds parameter missing from TileLayerOptions

### ISSUE THREAD

**Issue by atiqsamtia** [fleaflet/flutter_map#1137]
Leaflet Js does provide `bounds` parameter for the Tiles layers as well so that those layers are only loaded for a certain area. just like bounds on the full map but this applies to individual layers. It is very good for having multiple layers from different locations active. Like separate maps for each area. 

DOCS:
https://leafletjs.com/reference.html#tilelayer-bounds

this implementation does not have this feature. Can you please provide this or guide me to implement it on my own? 

![image](https://user-images.githubusercontent.com/5112753/150660376-c1c219ac-3b7b-4c6e-8cc1-cb8ccaf1bb94.png)

**Comment by ibrierley** [comment]
Have you tried 

```
MapOptions(
  bounds: LatLngBounds(LatLng(58.8, 6.1), LatLng(59, 6.2)),
  boundsOptions: FitBoundsOptions(padding: EdgeInsets.all(8.0)),
),
```

**Comment by atiqsamtia** [comment]
@ibrierley Yes, I've used this one. but this is for a full map. The missing one is for each layer. 

So that if multiple layers are added on the map, unnecessary 404 tiles requests are not sent if map is moved out of specific layer bounds 

leaflet JS supports this natively.

**Comment by hpoul** [comment]
Is there any workaround for this? I guess I could use a custom TileProvider which doesn't request tiles which are out of bounds? 🤔 I also have a custom map overlay which is only available for a certain region, and everything outside of that map i want OSM to be displayed.. Essentially everything works, except generating a ton of 404s :) so it would be nice to have a bounds for a single tile layer

**Comment by ibrierley** [comment]
Just trying to understand this a bit more...if the mapoptions has a bounds set, the tilelayer shouldn't request other tiles outside those bounds should it ? Or are you saying the tile provider is grabbing tiles out of the bounds area ?

Trying to be clear on the difference, and why the tileprovider is requesting unnecessary tiles...

**Comment by hpoul** [comment]
@ibrierley Not sure I understand, I think the MapOptions are for the whole `FlutterMap`, or can I set a different `MapOptions` for each layer? I have an OpenStreet Map tile layer as base map for the whole map, but want to have my own map tiles overlayed for a single county / region. So I have one `TileLayerWidget` with OSM for the whole world.. and my custom `TileLayerWidget` which should only be requested for a smaller bound. Does this make sense?

**Comment by ibrierley** [comment]
Yes, the bounds option is for a smaller region of the whole map, that you can't navigate over iirc...so I'm a bit unclear why one wants different bounds for different layers...I'm not saying there isn't a use case for it...just trying to understand what the difference in display and tile requests would be.

So you do mean you would for example display a layer of full openstreetmaps, with no bounds (so osm when navigating the USA), and then UK specific adapted tiles with bounds on the layer above when navigating the UK for example ?

**Comment by hpoul** [comment]
> So you do mean you would for example display a layer of full openstreetmaps, with no bounds (so osm when navigating the USA), and then UK specific adapted tiles with bounds on the layer above when navigating the UK for example ?

Exactly, in my case we have our own branded, stylized version of a map highlighting specific routes. But I still want to allow the user to navigate outside of that region.

<img width="1196" alt="Screen Shot 2022-04-07 at 10 36 46" src="https://user-images.githubusercontent.com/313066/162157579-43e9781c-9558-4acc-a9fb-5b9dba9f9fcc.png">

**Comment by atiqsamtia** [comment]
Another use-case for TileLayer specific bounds is when we have multiple layers for example One layer for London and One layer for Paris. when a user is in the Paris area, the London layer should not request tiles as it'll return a massive 404 and vice versa. 

Leaflet has this built-in in JS version. 

I wanted to send a PR but there is no documentation on how to add a new TileLayer and how the data is flowing, I tried but could not get the bounds. 

I suggest creating a new `BoundingTileLayer` so that other users don't get affected and normal layers don't run the bonding calculation for each tile.

**Comment by hpoul** [comment]
> wondering if it would make sense of have a bounds check in the tile_layer.dart around lines 1002

I guess this sounds reasonable.. currently `_isValidTile` just checks for `globalBounds` I guess it can't be hard to add a `layerBounds` or something. Maybe i'll give it a go when i've got time

**Comment by ibrierley** [comment]
Thinking about this a bit more...is it enough for just the tiles not to be loaded that are outside these bounds...or should it be a specific display thing ? 
I'm guessing it may be easier to not include tiles outside of certain bounds, than needing a display and maybe needing new sized widgets or something to be added ?

**Comment by hpoul** [comment]
@ibrierley for my usecase it's just about the loading..  the current behavior from a user's point of view is already perfect, i just want to get rid of those 404s 😅

**Comment by ibrierley** [comment]
do you have a small example highlighting the problem to reproduce ?

**Comment by ibrierley** [comment]
Sorry, I didn't mean it was unclear, I was just trying to figure a useful test, but realised I only needed one layer to test anyway....

I've created a PR https://github.com/fleaflet/flutter_map/pull/1212   give that a go, see if it does what you want before we consider merging it.

You can add tileBounds: LatLngBounds(LatLng(43,19), LatLng(48,27)), as an example for Romania

**Comment by JaffaKetchup** [comment]
Closing, as fixed by #1212.
